File: server/server/authentication/services/authentication/authentication_service_impl.py
from typing import Optional

from django.contrib.auth.models import User

from server.authentication.services.authentication.authentication_service import IAuthenticationService
from server.authentication.services.authentication.errors import InvalidLoginException
from server.authentication.services.authentication.params import AuthenticateAccountParams


class AuthenticationServiceImpl(IAuthenticationService):
    def authenticate_account(self, params: AuthenticateAccountParams) -> User:
        account: Optional[User] = User.objects.filter(username=params.username).first()

        if not account or not account.check_password(params.password) or not account.is_active:
            raise InvalidLoginException()

        return account

File: client/lib/features/authentication/domain/services/authentication_service_impl.dart
import 'package:client/features/authentication/domain/exceptions/invalid_login_exception.dart';
import 'package:client/features/authentication/domain/models/application_user.dart';
import 'package:client/features/authentication/domain/services/authentication_service.dart';
import 'package:client/integrations/server_sdk.dart';
import 'package:server_api_client_dart/server_api_client_dart.dart';

class AuthenticationServiceImpl implements IAuthenticationService {
  const AuthenticationServiceImpl();

  @override
  Future<ApplicationUser> authenticate(String username, String password) async {
    // Perform the authentication with the server client
    final AuthenticateIdentityResponse response;
    try {
      response = (await ServerClient.instance.getIdentityApi().authenticate(
          authenticateIdentityRequest: (AuthenticateIdentityRequestBuilder()
            ..username = username
            ..password = password
          ).build()
      )).data!;
    } on ServerClientError catch(ex) {
      throw InvalidLoginException(ex.message);
    }

    ServerClient.instance.setAccessToken(response.access);

    return ApplicationUser(
      id: response.account.pk,
      username: response.account.username,
      created: response.account.dateJoined ?? DateTime.now(),
      updated: response.account.lastLogin ?? DateTime.now(),
    );
  }

  @override
  Future<void> deauthenticate() async {
    ServerClient.instance.clearAccessToken();
  }
}
